Dai intel dmic additions

juimonen opened the pull request:
- add dmic dts definitions for ace platform
- add support in the code for tgl and apl platforms (add cavs25 and 15 dts dmic defs)
- misc dmic code fixes found in sof native driver compilation

Commit:
drivers: dai: intel: dmic: misc fixes and changes

Fix misc issues and warnings found with sof native drivers compilation.

Signed-off-by: Jaska Uimonen <jaska.uimonen@linux.intel.com>

Commit:
dts: xtensa: intel: add dmic dts definitions

Add Intel dmic dai dts definitions for ace15, cavs15 and cavs25
platforms.

Add also pre dts cmake files as the dmic nodes use same address and emit
warnings otherwise.

Signed-off-by: Jaska Uimonen <jaska.uimonen@linux.intel.com>

andyross commented on dts/xtensa/intel/intel_cavs15.dtsi:
```diff
@@ -281,5 +281,22 @@
+		       interrupt-parent = <&cavs3>;
+		};
+
+		dmic1: dai-dmic@10001 {
```
same note

andyross commented on dts/xtensa/intel/intel_cavs25.dtsi:
```diff
@@ -265,5 +265,21 @@
+		       interrupt-parent = <&cavs3>;
+		};
+
+		dmic1: dai-dmic@10001 {
```
ditto

andyross commented on drivers/dai/intel/dmic/dmic.c:
```diff
@@ -31,6 +31,82 @@ struct dai_dmic_global_shared dai_dmic_global;
+ * The input is Q3.29
+ * The output is Q9.23
+ */
+static int32_t exp_small_fixed(int32_t x)
```
I know this isn't really related to this PR and you're just moving this around, but this code is weirding me out.  The DMIC driver should be copying raw data from the device to an application layer.  If there is rescaling/transformation/resampling/whatever needed that really should be done at the app layer.  Zephyr doesn't (and shouldn't!) know anything about how PDM microphone signal processing works, that's what SOF is for.

juimonen replied:
I think the original reason to have it very close to hardware is that there's a "feature" in dmic hw, that makes a "pop" (unwanted spike in the waveform) in the very beginning. The dmic does some processing in hw ip, like decimation filtering etc. but there's no hw implementation for ramp or any other algorithm to get rid of the error.

So we could have a "ramp component" in sof pipeline, but what it would essentially do is to make a couple of millisecond ramp, then just pass through the 99,9% of the samples.

In this PR I was just moving this away from the header as it was giving me warning from dmic.nhlt.c

andyross reviewed (commented):
That address trick seems like an abuse of DTS (to my somewhat amateur eyes) and there should be cleaner ways to do what you want (including just using the DT_INST index, since you control the order in the dts file).  @galak @mbolivar might have better suggestions.

juimonen commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -277,6 +277,22 @@
+			interrupt-parent = <&ace_intc>;
+		};
+
+		dmic1: dai-dmic@10001 {
```
yes correct. I just basically want to have 2 entries in dts to be able to instantiate 2 separate dmic dais. In practice it is the 1 and same hw block, but it can operate 2 streams at the same time (we can have for example 48kHz and 16kHz streams from same microphones at the same time).

Historically in sof these have been treated as separate dais, which I still think is the most clear thing to do. People start to get really confused when starting to talk about dmic internal pdm and fifo structures.

If someone has better suggestion, I will change this for sure.

juimonen replied:
@galak @mbolivar @ceolin any suggestions how to do this nicely the "correct way"? see my explanation above.

juimonen replied:
@andyross tried to "clean" the dts by adding separating fifo field into the yaml.... not sure this is better/cleaner.

Commit:
dts: bindings: add fifo and modify shim in intel,dai-dmic

Add a fifo address as int value into dmic yaml to separate different
outputs (corresponds to different dais) from the same hw block. Also
change shim address from array to int value.

Signed-off-by: Jaska Uimonen <jaska.uimonen@linux.intel.com>

Commit:
drivers: dai: intel: dmic: use fifo address from devicetree

Use fifo address from device tree to create the device structs.
Also modify shim initialization as it was changed from array to
int. Fix cosmetic tab in device init macro.

Signed-off-by: Jaska Uimonen <jaska.uimonen@linux.intel.com>

andyross commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -277,6 +277,24 @@
+		dmic0: dai-dmic@10000 {
+			compatible = "intel,dai-dmic";
+			reg = <0x10000 0x8000>;
+			fifo = <0x0008 0x100>;
```
This seems reasonable to me (again: with "me" **not** being the best reference for DTS design patterns).  But AFAICT you're only ever using the first entry of this array in the driver code.  What's does the second value mean?  Also, is there a way to construct this from a simpler "index" instead of a raw register offset?  Maybe the hardware has microphone register blocks in an array with a stride of 256 bytes?

ceolin replied:
Yep, what does the second value means ?

juimonen replied:
changed now to int value

ceolin commented on drivers/dai/intel/dmic/dmic.c:
```diff
@@ -174,30 +250,55 @@ static void dai_dmic_irq_handler(const void *data)
 static inline void dai_dmic_dis_clk_gating(const struct dai_intel_dmic *dmic)
 {
+#if CONFIG_SOC_SERIES_INTEL_CAVS_V15
+	(void)index;
```
where this variable comes from ?

juimonen replied:
copy paste error from sof code... removing

ceolin commented on drivers/dai/intel/dmic/dmic.c:
```diff
@@ -174,30 +250,55 @@ static void dai_dmic_irq_handler(const void *data)
 static inline void dai_dmic_en_clk_gating(const struct dai_intel_dmic *dmic)
 {
+#if CONFIG_SOC_SERIES_INTEL_CAVS_V15
+	(void)index;
```
ditto

andyross commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -277,6 +277,24 @@
+		dmic0: dai-dmic@10000 {
+			compatible = "intel,dai-dmic";
+			reg = <0x10000 0x8000>;
+			fifo = <0x0008>;
```
Still nitpicky: you dropped the second item, but now it's an array of one element.  You can have integer properties ("type: int"), which seems like what you really want.

Also I still think this would be clearer if it stored an index (0 or 1) instead of a register offset.

(Not a -1, just a note)

juimonen replied:
np, good to get feedback. I just noticed that zephyr is silently merging the 2 nodes with same register address.. ??? So I don't get 2 devices, only 1. So I probably need to come up with some other scheme here for the nodes.

mbolivar-nordic replied:
> So I probably need to come up with some other scheme here for the nodes.

Assuming that they really are two different peripherals that just happen to have the same register map addresses, you can just give them different names. Examples:

https://github.com/zephyrproject-rtos/zephyr/blob/6f9e8f459d261f36dc4cf7e50376f4a4931e39ea/dts/arm/nordic/nrf52840.dtsi#L87
https://github.com/zephyrproject-rtos/zephyr/blob/6f9e8f459d261f36dc4cf7e50376f4a4931e39ea/dts/arm/nordic/nrf52840.dtsi#L104

This will cause a warning which you can disable with https://github.com/zephyrproject-rtos/zephyr/blob/6f9e8f459d261f36dc4cf7e50376f4a4931e39ea/boards/arm/nrf52840dk_nrf52840/pre_dt_board.cmake#L7.

mbolivar-nordic replied:
... although you did just find a bug in dtlib. I assumed `dtc` would have the same behavior, but it actually treats this as an error by default:

```
$ cat foo.dts
/dts-v1/;

/ {
	blub: foo@1234 {
		reg = <1234 4>;
		bar = <1>;
	};

	blab: foo@1234 {
		reg = <1234 4>;
		baz = <2>;
	};
};

$ dtc -Odts foo.dts
foo.dts:8.17-10.4: ERROR (duplicate_node_names): /foo@1234: Duplicate node name
ERROR: Input tree has errors, aborting (use -f to force output)
```

We are compatible with dtc by contract, so this should have been an error. I'll file a bug, thank you for noticing this.

juimonen replied:
@mbolivar-nordic I tried to do this as you suggested... hopefully kind of what you were asking

mbolivar-nordic commented on drivers/dai/intel/dmic/dmic.c:
```diff
@@ -785,7 +785,7 @@ static int dai_dmic_initialize_device(const struct device *dev)
 		{							\
 			.offset = DT_INST_REG_ADDR_BY_IDX(n, 0)		\
-				+ OUTDATA##n,				\
+				+ DT_INST_PROP_BY_IDX(n, fifo, 0),	\
```
I think I agree with @andyross; this should be an int if you only expect one value inside. Unless there's something I'm missing? The commit logs don't have much information and there is no property description documenting what this is supposed to contain, so I'm going to -1 until we can clear this up.

If you switch to int, this just becomes `DT_INST_PROP(n, fifo)`.

juimonen replied:
ok, should be like that now

mbolivar-nordic replied:
>  The commit logs don't have much information and there is no property description documenting what this is supposed to contain, so I'm going to -1 until we can clear this up.

This was a hint to improve the commit log and add a property description, just FYI. I'm going to dismiss my review now because it's moved to `int`, but I think these should be improved.

Commit:
drivers: dai: intel: dmic: add cavs15 and 25 support

Add support for older platforms like cavs15 and 25 into intel dmic dai.

Signed-off-by: Jaska Uimonen <jaska.uimonen@linux.intel.com>

juimonen commented on drivers/dai/intel/dmic/dmic.c:
```diff
@@ -174,30 +174,53 @@ static void dai_dmic_irq_handler(const void *data)
 
 static inline void dai_dmic_dis_clk_gating(const struct dai_intel_dmic *dmic)
 {
+#if CONFIG_SOC_SERIES_INTEL_CAVS_V15
```
ok, fixed now

juimonen commented:
@nashif can't really is this PR causing the unit test failures...? seems unrelated.

juimonen commented:
@abonislawski @marcinszkudlinski @dcpleung @kv2019i  sorry to spam, but rebased now and CI clean. So need re-approvers if found ok :) thanks

kv2019i reviewed (dismissed):
Thanks @juimonen , looks good!

jxstelter commented on drivers/dai/intel/dmic/dmic.c:
```diff
@@ -174,30 +250,53 @@ static void dai_dmic_irq_handler(const void *data)
 
 static inline void dai_dmic_dis_clk_gating(const struct dai_intel_dmic *dmic)
 {
+#ifdef CONFIG_SOC_SERIES_INTEL_CAVS_V15
```
I think it would be good to have shim registers definitions and basic access functions separated from main DMIC driver code and moved to platform specific headers. This patch now adds #ifdef CONFIG_SOC_SERIES_INTEL_CAVS_V15, but shim registers definition for SOC_SERIES_INTEL_ACE will change from one version to another.  In result soon we would have here several #ifdefs  for different generations.

kv2019i replied:
@juimonen @jxstelter Is this something to do now, or later in separate PR. Given we only have two cases now, I'd prefer to do this in separate PR.

jxstelter replied:
OK.  Separate PR will be good.

dcpleung commented:
Could you rebase to get rid of the merge commit?

juimonen commented:
> Could you rebase to get rid of the merge commit?

yeah sorry, I think I accidentally pushed the update branch button here in github...
